cloudsim-plus/src/main/java/org/cloudsimplus/faultinjection/VmCloner.java

```diff
@@ -0,0 +1,98 @@
+package org.cloudsimplus.faultinjection;
+
+import org.cloudbus.cloudsim.cloudlets.Cloudlet;
+import org.cloudbus.cloudsim.vms.Vm;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+
+/**
+ * Enables cloning a {@link Vm} which was destroyed
+ * due to a {@link HostFaultInjection Host Failure}.
+ * It provides all the features to clone a Vm, simulating
+ * the creating of another Vm from an snapshot of the failed one.
+ * It also enables re-creating Cloudlets which were running
+ * inside the failed VM.
+ *
+ * @author raysaoliveira
+ * @since CloudSim Plus 1.2.2
+ */
+public interface VmCloner {
+    VmCloner NULL = new VmCloner() {
+        @Override public int getClonedVmsNumber() { return 0;}
+        @Override public Map.Entry<Vm, List<Cloudlet>> clone(Vm sourceVm) { return new HashMap.SimpleEntry<>(Vm.NULL, Collections.EMPTY_LIST); }
+        @Override public VmCloner setVmClonerFunction(UnaryOperator<Vm> vmClonerFunction) { return this; }
+        @Override public VmCloner setCloudletsClonerFunction(Function<Vm, List<Cloudlet>> cloudletsClonerFunction) { return this; }
+        @Override public int getMaxClonesNumber() { return 0; }
+        @Override public boolean isMaxClonesNumberReached() { return false; }
+        @Override public VmCloner setMaxClonesNumber(int maxClonesNumber) { return this; }
+    };
+
+    /**
+     * Gets the number of VMs cloned so far.
+     * @return
+     */
+    int getClonedVmsNumber();
+
+    /**
+     * Clones a given {@link Vm} using the Vm Cloner Function
+     * and their Cloudlets using the Clodlets Cloner Function,
+     * binding the cloned Cloudlets to the cloned Vm.
+     *
+     * @param sourceVm the Vm to be cloned
+     * @return a {@link Map.Entry} where the key is the cloned Vm
+     * and the value is the List of cloned Cloudltes.
+     * @see #setVmClonerFunction(UnaryOperator)
+     * @see #setCloudletsClonerFunction(Function)
+     */
+    Map.Entry<Vm, List<Cloudlet>> clone(Vm sourceVm);
+
+    /**
+     * Sets the {@link UnaryOperator} to be used to clone {@link Vm}s.
+     * It is a Function which, when called, creates a clone of a specific Vm.
+     * @param vmClonerFunction the {@link Vm} cloner Function to set
+     * @return
+     */
+    VmCloner setVmClonerFunction(UnaryOperator<Vm> vmClonerFunction);
+
+    /**
+     * Gets the {@link Function} to be used to clone Vm's {@link Cloudlet}s.
+     * It is a Function which, when called, creates a clone of cloudlets
```

Comment:
Change "It is a Function which, when called," to "When the given Function is called,"